TypeSystem/Reflection/schema.h

```diff
@@ -244,107 +297,106 @@ struct StructSchema {
+            return DescribeOptional(language, cit->second);
+          }
+          return "UNHANDLED_TYPE_" + bricks::strings::ToString(type_id);
+        }
       }
-      if (type_prefix == TYPEID_ENUM_PREFIX) {
-        return schema_.types[type_id].enum_name;
-      }
-      if (type_prefix == TYPEID_VECTOR_PREFIX) {
-        return DescribeCppVector(schema_.types[type_id]);
-      }
-      if (type_prefix == TYPEID_PAIR_PREFIX) {
-        return DescribeCppPair(schema_.types[type_id]);
-      }
-      if (type_prefix == TYPEID_MAP_PREFIX) {
-        return DescribeCppMap(schema_.types[type_id]);
-      }
-      if (type_prefix == TYPEID_OPTIONAL_PREFIX) {
-        return DescribeCppOptional(schema_.types[type_id]);
-      }
-      return "UNHANDLED_TYPE_" + bricks::strings::ToString(type_id);
     }
   }
 
-  std::string DescribeCppVector(const TypeInfo& type_info) {
+  std::string DescribeVector(const Language::CPP& language, const TypeInfo& type_info) const {
+    assert(type_info.included_types.size() == 1u);
+    return "std::vector<" + TypePrintName(language, type_info.included_types[0]) + '>';
+  }
+  std::string DescribeVector(const Language::FSharp& language, const TypeInfo& type_info) const {
     assert(type_info.included_types.size() == 1u);
-    return "std::vector<" + CppType(type_info.included_types[0]) + '>';
+    return TypePrintName(language, type_info.included_types[0]) + " array";
   }
 
-  std::string DescribeCppPair(const TypeInfo& type_info) {
+  std::string DescribePair(const Language::CPP& language, const TypeInfo& type_info) const {
     assert(type_info.included_types.size() == 2u);
-    return "std::pair<" + CppType(type_info.included_types[0]) + ", " + CppType(type_info.included_types[1]) +
-           '>';
+    return "std::pair<" + TypePrintName(language, type_info.included_types[0]) + ", " +
+           TypePrintName(language, type_info.included_types[1]) + '>';
+  }
+  std::string DescribePair(const Language::FSharp& language, const TypeInfo& type_info) const {
+    assert(type_info.included_types.size() == 2u);
+    return TypePrintName(language, type_info.included_types[0]) + " * " +
+           TypePrintName(language, type_info.included_types[1]);
   }
 
-  std::string DescribeCppMap(const TypeInfo& type_info) {
+  std::string DescribeMap(const Language::CPP& language, const TypeInfo& type_info) const {
+    assert(type_info.included_types.size() == 2u);
+    return "std::map<" + TypePrintName(language, type_info.included_types[0]) + ", " +
+           TypePrintName(language, type_info.included_types[1]) + '>';
+  }
+  std::string DescribeMap(const Language::FSharp& language, const TypeInfo& type_info) const {
     assert(type_info.included_types.size() == 2u);
-    return "std::map<" + CppType(type_info.included_types[0]) + ", " + CppType(type_info.included_types[1]) +
-           '>';
+    return "UNSUPPORTED_MAP<" + TypePrintName(language, type_info.included_types[0]) + ", " +
```

Comment:
Never tested it within a discriminated union (may need `new` of some sort). Let's leave it as is now, our F# bindings would not need any per-entry dictionaries yet. Adding a `TODO(dkorolev)`.